### DISCUSSION TITLE
Provide Nix shell environment

### DISCUSSION THREAD

**Discussion by koslambrou** [CardanoSolutions/kupo#113] [Ideas]
### What is your idea? Provide a use case.

When wanting to develop on Kupo, I would be nice to have a `shell.nix` file enabling a development environment using `nix-shell` (or `nix develop` is using flakes). Could use `haskell.nix`.

### Why is it a good idea?

Because just running `nix-shell` and having all the development environment setup is a good idea.

### What is the current alternative and why is it not good enough?

Currentl alternative is installing dependencies ourselves and making sure we use the correct versions (GHC, cabal, haskell-language-server, etc.)

**Comment by KtorZ** [comment]
> Current alternative is installing dependencies ourselves

I like this alternative to be honest 😶 ... 

More seriously, I am generally quite opposed to Nix for deployment and package management in projects because it has never solved an actual problem for me -- yet has always added complexity and frustration. However, I am open to the idea of using Nix for provisioning environments. So a `shell.nix` that fulfill this role could be nice. 

I have seldom used `haskell.nix` (only for producing static binaries) and it's not something I enjoy doing as I am not a Nix user. I don't mind a good old system-level installation. Hence, if you're willing to provide a PR with a `shell.nix` that can be used to provision an environment, I'll review it and see about merging it. 

Having said that:

- It should not require generating or re-generating some random lock files. This is a nightmare to maintain down the line then.
- It should not use Nix to fetch Haskell dependencies; I still want plain cabal to handle the Haskell build. So I am fine with using Nix to provision the dev environment, but nothing more. 

Perhaps something worth looking at is what Moritz has been up to with: https://github.com/input-output-hk/devx. I've been playing around that a bit but put that on hold for now. Yet I think he's doing exactly that.
